src/rust/engine/process_execution/src/local.rs

```diff
@@ -275,6 +278,27 @@ impl super::CommandRunner for CommandRunner {
   }
 }
 
+use ouroboros::self_referencing;
+
+#[self_referencing]
+struct OwnedChild {
+  child: Box<Child>,
+  #[borrows(mut child)]
+  #[not_covariant]
+  exit_stream: BoxStream<'this, Result<ChildOutput, std::io::Error>>,
+}
+
+impl Stream for OwnedChild {
+  type Item = Result<ChildOutput, std::io::Error>;
+
+  fn poll_next(
+    mut self: Pin<&mut Self>,
+    cx: &mut std::task::Context<'_>,
+  ) -> Poll<Option<Self::Item>> {
+    self.with_exit_stream_mut(|es| Pin::new(es).poll_next(cx))
+  }
+}
+
```

Comment:
`tokio::process::Child` no longer implements `Future` in Tokio 1.0. Instead, `Child` exposes a `wait` instance method. The problem is that `wait` takes `&mut self` and so the original version of `exit_stream` thus mutably borrows `self` and so cannot be returned from `run_in_workdir` since the borrow checker rejects the code since `Child` does not survive the call to `run_in_workdir`.

The solution is to have `Child` survive that call. This code uses the `ouroboros` crate which has a proc macro that allows for a limited form of self-referential structs. The `Child` struct is owned by this struct and this struct is what comprises part of the stream returned from `run_in_workdir`. The `exit_stream` part of the return value from `run_in_workdir` is also embedded in this struct using the `ouroboros` crate.

Finally, the solution implements `Stream` on `OwnedChild` so that it exposes the same stream as `exit_stream` but as a first-class owner of the embedded `Child`.

cc @stuhood

Reply:
So, afaik, the suggested way to do this is to call `child.stdout.take()` (which is identical to how you do this with the `stdlib` `Child` type). See the `stdout`/`stderr` fields in the stdlib docs https://doc.rust-lang.org/std/process/struct.Child.html for more info. You can then call `wait` safely.

Reply:
The issue is not with `child.stdout` nor with `child.stderr`. The `stdout_stream` and `stderr_stream` streams take ownership of those respective values and are not triggering the borrow checker. See the construction of `stdout_stream` and `stderr_stream` where the code already calls `child.stdout.take().unwrap()` and `child.stderr.take().unwrap()` respectively.

The issue is with `exit_stream`. In tokio v0.2.x, `Child` implements `Future` and can be used directly to await the exit status of the child process. The borrow checker was not triggered in `run_in_workdir` previously because `exit_stream` took ownership of the underlying `Future` a/k/a the `Child` struct. Thus returning `exit_stream` from `run_in_workdir` implicitly passed ownership of `Child` out of `run_in_workdir` because `exit_stream` owned it.

In tokio v1.x, that is not the case, `Child` no longer implements `Future` and one must call `.wait` to obtain a Future that can be used to await the child's exit status. `.wait` takes `&mut self` and thus borrows `Child` mutably within the call to `run_in_workdir`. Unlike in tokio v0.2.x, the borrow checker triggers in this case because `exit_stream` now only has the `&mut self` reference to `Child` (via `.wait`) and not ownership of `Child` like it did previously. Thus the borrow checker triggers because the `Child` is dropped when `run_in_workdir` returns.

The solution is to let `Child` escape `run_in_workdir`. However, the code should also keep the Future returned from `.wait` alive (and its `&mut self` reference) so that we only construct `exit_stream` once. This is accomplished by using the `ouroboros` crate to safely maintain the self-reference between `Child` and `exit_stream`.

And to make things easier, the `OwnedChild` struct implements Stream so that the details are hidden of how `exit_stream` and the `Child` struct are stored.

Reply:
Hm. Ok. I expect that moving forward we should move away from the "unified stream of stdout and exit code" interface here, as TODO'd in: https://github.com/pantsbuild/pants/pull/11370/files#r547623793. `nails` did to ease lifetime issues, and it seemed to work fine. Happy to land this as is if we preserve the TODO though.

Reply:
Fine by me. I'll preserve the TODO.